File: sdks/python/linktor/utils/webhook.py
"""
Webhook signature verification utilities
"""

import hashlib
import hmac
import json
import time
from typing import Any, Callable, Optional, TypeVar

from linktor.types.webhook import WebhookEvent, WebhookEventType

SIGNATURE_HEADER = "x-linktor-signature"
TIMESTAMP_HEADER = "x-linktor-timestamp"
DEFAULT_TOLERANCE = 300  # 5 minutes

T = TypeVar("T")


def compute_signature(payload: str | bytes, secret: str) -> str:
    """
    Compute HMAC-SHA256 signature

    Args:
        payload: Raw request body
        secret: Webhook secret

    Returns:
        Hex-encoded signature
    """
    if isinstance(payload, str):
        payload = payload.encode("utf-8")

    return hmac.new(
        secret.encode("utf-8"),
        payload,
        hashlib.sha256,
    ).hexdigest()


def verify_webhook_signature(
    payload: str | bytes,
    signature: str,
    secret: str,
) -> bool:
    """
    Verify webhook signature

    Args:
        payload: Raw request body
        signature: Signature from x-linktor-signature header
        secret: Webhook secret

    Returns:
        True if signature is valid
    """
    if not signature or not secret:
        return False

    expected_signature = compute_signature(payload, secret)

    # Timing-safe comparison
    return hmac.compare_digest(signature, expected_signature)


def verify_webhook(
    payload: str | bytes,
    headers: dict[str, str],
    secret: str,
    tolerance: int = DEFAULT_TOLERANCE,
) -> bool:
    """
    Verify webhook with timestamp validation

    Args:
        payload: Raw request body
        headers: Request headers (case-insensitive)
        secret: Webhook secret
        tolerance: Maximum age of the webhook in seconds

    Returns:
        True if signature and timestamp are valid
    """
    # Normalize header keys to lowercase
    headers_lower = {k.lower(): v for k, v in headers.items()}

    signature = headers_lower.get(SIGNATURE_HEADER)
    timestamp = headers_lower.get(TIMESTAMP_HEADER)

    if not signature:
        return False

    # Verify timestamp if present
    if timestamp:
        try:
            webhook_time = int(timestamp)
            now = int(time.time())
            if abs(now - webhook_time) > tolerance:
                return False
        except ValueError:
            return False

    return verify_webhook_signature(payload, signature, secret)


def construct_event(
    payload: str | bytes,
    headers: dict[str, str],
    secret: str,
    tolerance: int = DEFAULT_TOLERANCE,
) -> WebhookEvent[Any]:
    """
    Parse and validate webhook event

    Args:
        payload: Raw request body
        headers: Request headers
        secret: Webhook secret
        tolerance: Maximum age in seconds

    Returns:
        Parsed webhook event

    Raises:
        ValueError: If verification fails or payload is invalid
    """
    if not verify_webhook(payload, headers, secret, tolerance):
        raise ValueError("Webhook signature verification failed")

    if isinstance(payload, bytes):
        payload = payload.decode("utf-8")

    try:
        data = json.loads(payload)
    except json.JSONDecodeError as e:
        raise ValueError(f"Invalid JSON payload: {e}")

    if not data.get("id") or not data.get("type") or not data.get("timestamp"):
        raise ValueError("Invalid webhook event structure")

    return WebhookEvent(**data)


def is_event_type(event: WebhookEvent[Any], event_type: WebhookEventType) -> bool:
    """
    Check if event is of a specific type

    Args:
        event: Webhook event
        event_type: Expected event type

    Returns:
        True if event matches the type
    """
    return event.type == event_type


WebhookHandler = Callable[[WebhookEvent[Any]], None]


def create_webhook_handler(
    secret: str,
    handlers: dict[WebhookEventType, WebhookHandler],
    tolerance: int = DEFAULT_TOLERANCE,
) -> Callable[[str | bytes, dict[str, str]], tuple[int, Optional[str]]]:
    """
    Create a webhook handler factory

    Args:
        secret: Webhook secret
        handlers: Dict mapping event types to handler functions
        tolerance: Maximum age in seconds

    Returns:
        Handler function that returns (status_code, body)
    """

    def handle_webhook(
        payload: str | bytes,
        headers: dict[str, str],
    ) -> tuple[int, Optional[str]]:
        try:
            event = construct_event(payload, headers, secret, tolerance)

            handler = handlers.get(event.type)
            if handler:
                handler(event)

            return 200, None
        except ValueError as e:
            return 400, str(e)
        except Exception as e:
            return 500, str(e)

    return handle_webhook


__all__ = [
    "compute_signature",
    "verify_webhook_signature",
    "verify_webhook",
    "construct_event",
    "is_event_type",
    "create_webhook_handler",
    "SIGNATURE_HEADER",
    "TIMESTAMP_HEADER",
    "DEFAULT_TOLERANCE",
]

File: sdks/typescript/src/utils/webhook.ts
/**
 * Webhook signature verification utilities
 */

import { createHmac, timingSafeEqual } from 'crypto';
import type { WebhookEvent, WebhookEventType } from '../types/webhook';

const SIGNATURE_HEADER = 'x-linktor-signature';
const TIMESTAMP_HEADER = 'x-linktor-timestamp';
const DEFAULT_TOLERANCE = 300; // 5 minutes

export interface WebhookVerificationOptions {
  /**
   * Maximum age of the webhook in seconds
   * @default 300 (5 minutes)
   */
  tolerance?: number;
}

export interface WebhookHeaders {
  [key: string]: string | string[] | undefined;
}

/**
 * Verify webhook signature
 *
 * @param payload - Raw request body as string or Buffer
 * @param signature - Signature from x-linktor-signature header
 * @param secret - Webhook secret
 * @param options - Verification options
 * @returns true if signature is valid
 */
export function verifyWebhookSignature(
  payload: string | Buffer,
  signature: string,
  secret: string,
  options?: WebhookVerificationOptions
): boolean {
  if (!signature || !secret) {
    return false;
  }

  const payloadString = typeof payload === 'string' ? payload : payload.toString('utf8');

  // Compute expected signature
  const expectedSignature = computeSignature(payloadString, secret);

  // Timing-safe comparison
  try {
    const sigBuffer = Buffer.from(signature, 'hex');
    const expectedBuffer = Buffer.from(expectedSignature, 'hex');

    if (sigBuffer.length !== expectedBuffer.length) {
      return false;
    }

    return timingSafeEqual(sigBuffer, expectedBuffer);
  } catch {
    return false;
  }
}

/**
 * Verify webhook with timestamp validation
 *
 * @param payload - Raw request body
 * @param headers - Request headers
 * @param secret - Webhook secret
 * @param options - Verification options
 * @returns true if signature and timestamp are valid
 */
export function verifyWebhook(
  payload: string | Buffer,
  headers: WebhookHeaders,
  secret: string,
  options?: WebhookVerificationOptions
): boolean {
  const signature = getHeader(headers, SIGNATURE_HEADER);
  const timestamp = getHeader(headers, TIMESTAMP_HEADER);

  if (!signature) {
    return false;
  }

  // Verify timestamp if present
  if (timestamp) {
    const tolerance = options?.tolerance ?? DEFAULT_TOLERANCE;
    if (!verifyTimestamp(timestamp, tolerance)) {
      return false;
    }
  }

  return verifyWebhookSignature(payload, signature, secret, options);
}

/**
 * Compute HMAC-SHA256 signature
 */
export function computeSignature(payload: string, secret: string): string {
  return createHmac('sha256', secret).update(payload, 'utf8').digest('hex');
}

/**
 * Verify timestamp is within tolerance
 */
function verifyTimestamp(timestamp: string, toleranceSeconds: number): boolean {
  const webhookTime = parseInt(timestamp, 10);
  if (isNaN(webhookTime)) {
    return false;
  }

  const now = Math.floor(Date.now() / 1000);
  const diff = Math.abs(now - webhookTime);

  return diff <= toleranceSeconds;
}

/**
 * Get header value (handles arrays)
 */
function getHeader(headers: WebhookHeaders, name: string): string | undefined {
  const value = headers[name] || headers[name.toLowerCase()];
  if (Array.isArray(value)) {
    return value[0];
  }
  return value;
}

/**
 * Parse and validate webhook event
 *
 * @param payload - Raw request body
 * @param headers - Request headers
 * @param secret - Webhook secret
 * @param options - Verification options
 * @returns Parsed webhook event
 * @throws Error if verification fails or payload is invalid
 */
export function constructEvent<T = unknown>(
  payload: string | Buffer,
  headers: WebhookHeaders,
  secret: string,
  options?: WebhookVerificationOptions
): WebhookEvent<T> {
  if (!verifyWebhook(payload, headers, secret, options)) {
    throw new Error('Webhook signature verification failed');
  }

  const payloadString = typeof payload === 'string' ? payload : payload.toString('utf8');

  try {
    const event = JSON.parse(payloadString) as WebhookEvent<T>;

    if (!event.id || !event.type || !event.timestamp) {
      throw new Error('Invalid webhook event structure');
    }

    return event;
  } catch (error) {
    if (error instanceof SyntaxError) {
      throw new Error('Invalid JSON payload');
    }
    throw error;
  }
}

/**
 * Type guard for webhook event types
 */
export function isEventType<T extends WebhookEventType>(
  event: WebhookEvent,
  type: T
): event is WebhookEvent & { type: T } {
  return event.type === type;
}

/**
 * Create a webhook handler factory for Express-like frameworks
 */
export function createWebhookHandler(
  secret: string,
  handlers: Partial<Record<WebhookEventType, (event: WebhookEvent) => void | Promise<void>>>,
  options?: WebhookVerificationOptions
) {
  return async (req: {
    body: string | Buffer | Record<string, unknown>;
    headers: WebhookHeaders;
  }): Promise<{ status: number; body?: string }> => {
    try {
      // Get raw body
      let rawBody: string | Buffer;
      if (typeof req.body === 'string' || Buffer.isBuffer(req.body)) {
        rawBody = req.body;
      } else {
        rawBody = JSON.stringify(req.body);
      }

      const event = constructEvent(rawBody, req.headers, secret, options);

      const handler = handlers[event.type];
      if (handler) {
        await handler(event);
      }

      return { status: 200 };
    } catch (error) {
      const message = error instanceof Error ? error.message : 'Unknown error';
      return { status: 400, body: message };
    }
  };
}
